### PR TITLE
Keaton coalesce typing fix

### PR DESCRIPTION
**Pull request by keatoooon**
Converts coalesce's type handling to be equivalent to the decomposed case statement. Previously, this was not the case, which caused issues when we stopped decomposing it.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/fun/SqlCoalesceFunction.java
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlCoalesceFunction.java:66]: Is this change to force a common return type?

File: core/src/main/java/org/apache/calcite/sql/fun/SqlCoalesceFunction.java
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlCoalesceFunction.java:102]: This comment was unintuitive to me at first (but definitely correct). Can you add an example of why for posterity?

File: core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java
Comment (njriasan) [inline core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java]: Is this name supposed to be TODO?

File: core/src/test/java/org/apache/calcite/rex/RexProgramTest.java
Comment (njriasan) [inline core/src/test/java/org/apache/calcite/rex/RexProgramTest.java:521]: I would add a comment that 2 will be cast to a string.

**Review by njriasan — COMMENTED** [review]
I left some comments but overall this looks good. Which tests are still failing?

File: core/src/main/java/org/apache/calcite/sql/fun/SqlCoalesceFunction.java
Comment (keatoooon) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlCoalesceFunction.java:66]: No, this is used to infer the type of the arguments to the function. I added a comment to make it clearer.

**Review by njriasan — APPROVED** [review]
LGTM! If you could fill out a brief description before merging that would be great.

**Pull request merged**
